fs-cache/src/cache.rs

```diff
@@ -0,0 +1,60 @@
+use crate::memory_limited_storage::MemoryLimitedStorage;
+use data_error::Result;
+use std::path::Path;
+
+pub struct Cache<K, V> {
+    storage: MemoryLimitedStorage<K, V>,
+}
+
+impl<K, V> Cache<K, V>
+where
+    K: Ord
+        + Clone
+        + serde::Serialize
+        + serde::de::DeserializeOwned
+        + std::fmt::Display
+        + std::hash::Hash
+        + std::str::FromStr,
+    V: Clone + serde::Serialize + serde::de::DeserializeOwned,
+{
+    pub fn new(
+        label: String,
+        path: &Path,
+        max_memory_bytes: usize,
+    ) -> Result<Self> {
+        log::debug!(
+            "{} cache initialized with {} bytes limit",
+            label,
+            max_memory_bytes
+        );
+        Ok(Self {
+            storage: MemoryLimitedStorage::new(label, path, max_memory_bytes)?,
+        })
+    }
+
+    pub fn get(&mut self, key: &K) -> Option<V> {
+        let result = self.storage.get(key);
+        log::debug!(
+            "{} cache: get key={} -> found={}",
+            self.storage.label(),
+            key,
+            result.is_some()
+        );
+        result
+    }
+
+    pub fn set(&mut self, key: K, value: V) -> Result<()> {
+        // Check if value already exists
+        if self.storage.get(&key).is_some() {
+            log::debug!(
+                "{} cache: skip setting existing key={}",
+                self.storage.label(),
+                key
+            );
+            return Ok(());
+        }
+
+        log::debug!("{} cache: set key={}", self.storage.label(), key);
```

Comment:
This debug statement is misleading. It should be placed after the call to `self.storage.set()`

Reply:
@tareknaser I think it's better to put the trace before the action. If something crashes down the stack, we'll have the log telling us what we tried to do before the crash.

Reply:
That's fine. I don’t have a strong opinion on this, but I think it should be something like 'trying to set key "{}"' since right now it sounds like the value is already set when it’s not. Maybe I’m overthinking it.

Reply:
I think logging both (before setting and after) will be more beneficial. I have added it.